### ISSUE TITLE
BUG - sev-2 - All - Unexpectedly formatted errors freezes home screen for 6-42 seconds

### ISSUE THREAD

**Issue by TKDickson** [department-of-veterans-affairs/va-mobile-app#7746]
## What happened?

Logging in as karsyn on a branch with with the changes from #7485, I get a 502 error returned immediately on the claims call on the home screen (details in the collapsible JSON section below)

However, like #4762 and #4879, we're not handling it well on the front end -- currently, the home screen freezes and takes ~6 seconds (Android) or ~42 seconds (iOS) to finally respond to further actions by the user.

## Specs:

- Device:
- OS:
- User Account: karsyn
- Accessibility State:

## Bug Severity - BE SURE TO ADD THE SEVERITY LABEL
 
See [Bug Tracking](https://department-of-veterans-affairs.github.io/va-mobile-app/docs/QA#issue-severity) for details on severity levels

- Impact: High Low 
- Frequency: High Low 

- 3 - Low
- 2 - High
- 1 - Critical

## Screen shot(s) and additional information

<details>
  <summary>Full JSON response for services related to issue (expand/collapse)</summary>
  
HTTP/1.1 502 Bad Gateway
Date: Thu, 18 Jan 2024 16:26:28 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 446
Referrer-Policy: strict-origin-when-cross-origin
Vary: Accept, Origin
X-Content-Type-Options: nosniff
X-Custom-Response-Header: eks-staging
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN
X-Git-Sha: e4540bb778047d496f4dba5409bcb4c82db682fe
X-Github-Repository: https://github.com/department-of-veterans-affairs/vets-api
X-Permitted-Cross-Domain-Policies: none
X-Request-Id: 79fe18d1-a613-4908-8f80-4ff9135ec93b
X-Runtime: 1.054000
X-Xss-Protection: 0
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
X-Frame-Options: SAMEORIGIN
Connection: keep-alive

{"data":[],"meta":{"errors":[{"service":"appeals","errorDetails":[{"title":"Not found","detail":"Appeals data for a veteran with that SSN was not found","code":"CASEFLOWSTATUS404","source":"A veteran with that SSN was not found in our systems.","status":"404"}]},{"service":"claims","errorDetails":"Forbidden: User is not authorized for claims"}],"pagination":{"currentPage":1,"perPage":10,"totalPages":0,"totalEntries":0},"activeClaimsCount":0}}
</details>

## Ticket Checklist
- [ ] Steps to reproduce are defined
- [ ] Desired behavior is added
- [ ] Labels added (front-end, back-end, relevant feature, accessibility, etc)
- [ ] Estimate of 1 added (for front-end tickets)
- [ ] Added either to the relevant feature epic (if found during new feature implementation), or the [bug epic](https://app.zenhub.com/workspaces/va-mobile-product-view-610035bc5395bb000e62e529/issues/gh/department-of-veterans-affairs/va-mobile-app/5367)

**Comment by alexandec** [comment]
I added some logging in api.ts and the slice and got the following output:

```
[0]  LOG  api.ts: received response for GET /v0/claims-and-appeals-overview
[0]  LOG  api.ts: response.status is 502
 LOG  api.ts: Content-Type is application/json; charset=utf-8. awaiting response.json
************ PAUSE FOR A FEW SECONDS HERE ************
[0]  LOG  api.ts: got response.json {"data": [], "meta": {"activeClaimsCount": 0, "errors": [[Object], [Object]], "pagination": {"currentPage": 1, "perPage": 10, "totalEntries": 0, "totalPages": 0}}}
[0]  LOG  slice: error fetching /v0/claims-and-appeals-overview. dispatching finish event
[0]  DEBUG  logging analytics event vama_error {"callStack": "{\"data\":[],\"meta\":{\"errors\":[{\"service\":\"appeals\",\"errorDetails\":[{\"title\":\"Not found\",\"detail\":\"Appeals data for a veteran with that SSN was not found\",\"code\":\"CASEFLOWSTATUS404\",\"source\":\"A veteran with that SSN was not found in our systems.\",\"status\":\"404\"}]},{\"service\":\"claims\",\"errorDetails\":\"Forbidden: User is not authorized for claims\"}],\"pagination\":{\"currentPage\":1,\"perPage\":10,\"totalPages\":0,\"totalEntries\":0},\"activeClaimsCount\":0}}", "endpoint": "/v0/claims-and-appeals-overview", "errorMessage": "status: 502 is network error undefined", "errorName": "status: 502 is network error undefined", "statusCode": 502}
```

The issue is caused by the following line in api.ts:

```
json = await response.json()
```

It appears this blocks the main thread so no other execution is possible until the response.json() call completes. No interaction with the app is possible while execution is blocked. It's surprising this method is so slow since the response is not very large. It's also strange that we only see slowness with this particular /v0/claims-and-appeals-overview call with Karsyn. Of course it's possible other types of errors are causing the same issue and we don't know about it.

When we reach the offending line we already know we have an error because `response.status > 399`. We're only calling response.json() to get more details on the error. So a potential solution would be to change `json = await response.json()` to something like:

```
response.json().then((responseJson) => {
  json = responseJson
  // existing code that was after json = await response.json()
})
```

This approach would allow the main thread to continue execution so the app remains responsive while the response.json() call is running. However, it means we continue execution immediately and throw an error without waiting until we have all the information we want in the error. Control moves back into the slice which catches the error and calls logNonFatalErrorToFirebase with the the small amount of info available at the time.

I tried throwing separately inside the .then() function like this:

```
if (response.headers.get('Content-Type')?.startsWith('application/json')) {
  response.json().then((responseJson) => {
    json = responseJson
    const vamfBody = json?.errors?.[0].source?.vamfBody

    if (vamfBody) {
      // Handle vamfBody separately since JSON.stringify chokes on it
      json.errors[0].source.vamfBody = ''
      text = JSON.stringify(json)
      const escaped = vamfBody.replace(/\\/g, '\\\\').replace(/"/g, '\\"')
      text = text.replace('"vamfBody":""', `"vamfBody":"${escaped}"`)
    } else {
      text = JSON.stringify(json)
    }

    throw { status: status, endpoint, text, json }
  })
} else {
  text = await response.text()
  json = {}
  throw { status: response.status, endpoint, text, json }
}
```

But again, control returns to the slice immediately, and the error only gets thrown later, after the slice has already finished its work. Our existing error handling structure requires that we wait until we have all the error information to proceed. I think we'd need to alter the approach if we want to report errors asynchronously whenever the error JSON is ready.

For what it's worth, I don't think this a new issue, just one we haven't noticed before. I found you can also reproduce the issue by logging in as Karsyn and going to Benefits -> Claims.

**Comment by alexandec** [comment]
### Further investigation

Here is the json response for `GET /mobile/v0/claims-and-appeals-overview` with Karsyn on staging which causes response.json() to choke (I formatted the response body for easier reading):

```
HTTP/1.1 502 Bad Gateway
Date: Mon, 22 Jan 2024 16:50:30 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Referrer-Policy: strict-origin-when-cross-origin
Vary: Accept, Origin
X-Content-Type-Options: nosniff
X-Custom-Response-Header: eks-staging
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN
X-Git-Sha: 9f5b907cdeb0037f37a3aff157e73faab87c5564
X-Github-Repository: https://github.com/department-of-veterans-affairs/vets-api
X-Permitted-Cross-Domain-Policies: none
X-Request-Id: e342cfbc-714a-4afd-a6d4-97d1fe34fc29
X-Runtime: 0.609516
X-Xss-Protection: 0
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
X-Frame-Options: SAMEORIGIN
Content-Length: 446

{
	"data": [],
	"meta": {
		"errors": [{
			"service": "appeals",
			"errorDetails": [{
				"title": "Not found",
				"detail": "Appeals data for a veteran with that SSN was not found",
				"code": "CASEFLOWSTATUS404",
				"source": "A veteran with that SSN was not found in our systems.",
				"status": "404"
			}]
		}, {
			"service": "claims",
			"errorDetails": "Forbidden: User is not authorized for claims"
		}],
		"pagination": {
			"currentPage": 1,
			"perPage": 10,
			"totalPages": 0,
			"totalEntries": 0
		},
		"activeClaimsCount": 0
	}
}
```

Using Charles, I set a breakpoint on the `GET /mobile/v0/claims-and-appeals-overview` request. By deleting various parts of the request, I found that the long pause is due to this endpoint returning errors for two different services at once, where one of the services has an array of errorDetails. Here are the scenarios I examined, with pause length for each:

1. One service returns an error, simple string = no pause
2. One service returns an error which is a list of errors = short pause
3. Two services return errors, simple strings for both = short pause
4. Two services return errors, either of which is a list of errors = long pause

In this case we have scenario 4 which produces a long pause. Of course there shouldn't be a pause occuring for any of these scenarios which points to an issue in response.json() itself. These react native issues are similar to what we are seeing:

https://github.com/facebook/react-native/issues/6418
https://stackoverflow.com/questions/36499603/taking-too-much-time-to-call-promises-then-function/36501330

It's unclear if these issues have been resolved.

### JSON.parse issue?

This is not a JSON.parse issue because if I run

```
JSON.parse('{"data":[],"meta":{"errors":[{"service":"appeals","errorDetails":[{"title":"Not found","detail":"Appeals data for a veteran with that SSN was not found","code":"CASEFLOWSTATUS404","source":"A veteran with that SSN was not found in our systems.","status":"404"}]},{"service":"claims","errorDetails":"Forbidden: User is not authorized for claims"}],"pagination":{"currentPage":1,"perPage":10,"totalPages":0,"totalEntries":0},"activeClaimsCount":0}}')
```

That returns immediately with a valid result. So there's some other issue with response.json().

### Errors in production?

We could investigate whether this is happening in production by looking at analytics events with the following:

```
endpoint: /v0/claims-and-appeals-overview
statusCode: 502
callStack contains both "service": "appeals" and "service": "claims"
```

Then look through the matching errors to see if the error responses contain a list of errors instead of just a string.

**Comment by alexandec** [comment]
I tested with the latest upgrades in the `CU/5486-Dylan-TypescriptDependencies` branch, but sadly the issue is still present there.

**Comment by dumathane** [comment]
> The (hopefully) last step in this one was bugging Greta to get me analytics regarding this endpoint and error type specifically. They were very accommodating and found that in the last six months the SSN appeal error only happened 72 times for 60 unique users and only on two specific days (which indicates to me there may have been a server or backend specific issue, and not a user data one). Even on those days the error only reproduced for a maximum of 0.028% of users. This makes me feel fine leaving the ticket and hoping someone in react-native land finds a solution and not forcing ourselves to make api changes or system wide error handling ones for such a small subset of a problem.
therese.dickson given this data, how feel you? Would you feel okay shipping given this info? Maybe drop to a sev 3?

**Comment by TKDickson** [comment]
No one is actively working on this; moving back to Current Sprint Work as a sev-2 bug.

We've had likely two user reports of this, come up in DSVA slack ([first](https://dsva.slack.com/archives/C018V2JCWRJ/p1712323650986289) and [second](https://dsva.slack.com/archives/C018V2JCWRJ/p1712946688274359)).

**Comment by TKDickson** [comment]
Nah it's a sev-2. The analytics were for when the call/error were only happening on the claims page; when we moved them to the homescreen the frequency increased. And anyways, since it's a full app freeze immediately upon login, it doesn't matter the frequency to be sev-2.

@bischoffa this ticket was blocked by a backend ticket that finally moved through this week (the blocker was in the dependencies) so will get moving again.

**Comment by bischoffa** [comment]
@TKDickson @ajsarkar28 looks like the blocker was removed 2 weeks ago. Can this be prioritized for this sprint given its a Sev-2 bug and one that has been lingering for almost 4 months now?

**Comment by TKDickson** [comment]
@bischoffa this can be closed!

afaik, the backend ticket fixed the most prevalent error case (type) that was causing this to happen for users. we're adding individual loading time metrics in #8570 that would hopefully help us track down further errors (although I have, and I know this is shocking, a lot of ideas about what we could be doing to better track error stuff).

So we may run into further reports of this in the future, but those would be resolved by (ideally) also backend fixes, since the underlying RN issue doesn't seem to be going away anytime soon
